Fix Issue 17525 - std.algorithm.searching.skipOver should have a single argument with pred version

RazvanN7 opened the pull request.

RazvanN7 commented:
Any ideas on why circleci is failing? The line that is reported doesn't even call empty

wilzbach commented:
https://github.com/dlang/phobos/pull/5497

wilzbach commented on std/algorithm/searching.d:
```diff
@@ -3969,6 +3984,15 @@ if (is(typeof(binaryFun!pred(r1.front, r2.front))) &&
+    auto s4 = "\t\t\t";
+    assert(s2.skipOver!isWhite && s2 == "value");
+    assert(!s3.skipOver!isWhite);
+    assert(s4.skipOver!isWhite && s3.empty);
```
You need to import empty from std.range.primitives here (public tests need to be runnable standalone)

RazvanN7 commented:
@wilzbach Ah, didn't see that one. Thanks anyway.

wilzbach commented:
> Ah, didn't see that one

As this is now the second PR in a row, you should start a NG thread to remind people to tag their PRs. In theory the bot could do this as well (I don't recall why it doesn't, feel free to open an issue)

RazvanN7 commented:
@wilzbach should I keep this PR open or should I close it and review @JackStouffer 's ?

wilzbach commented:
> @wilzbach should I keep this PR open or should I close it and review @JackStouffer 's ?

As the other PR has a rather long discussion, I would first read through it and try to find out the consensus. IIRC there wasn't none yet. Once there is an agreement, you can always reopen this PR if necessary (or create a new one from the same branch). Sounds good?

wilzbach commented:
> Next time please add "Fix Issue $Number" so that the dlang bot can associate your PR with the bug fix. 

Btw do you see the `✗` on the bug list summary? It means the issue is just "mentioned" and won't be automatically closed.
I have opened a PR to clarify this -> https://github.com/dlang-bots/dlang-bot/pull/124

RazvanN7 commented:
@wilzbach Oh I get it now. I didn't thought at all what that tiny x means. I rebased and added the proper name. Thanks for all the enlightening comments!

RazvanN7 commented:
@andralex it's there: https://github.com/dlang/phobos/pull/5543/files#diff-8b73351c2fbfb7fef513b00ced46d022R3989 , but I amended the commit so that there is no need to squash

CyberShadow commented:
Err, I'm certain I must have pointed out somewhere (not this thread I guess) that this implementation is inconsistent with the other `skipOver` overloads. This one loops, whereas the others don't.

@andralex Does this make sense to you?

RazvanN7 commented:
@CyberShadow What do you mean by loop? https://github.com/dlang/phobos/pull/5543/files#diff-8b73351c2fbfb7fef513b00ced46d022R3947 : this seems like a loop to me

CyberShadow commented:
That loop simply consumes the two input ranges for comparison; it does not consume multiple instances of the "prefix". Consider the difference between `s.skipOver(' ')` and `s.skipOver!(c => c == ' ')` - intuitively one may think that these should behave the same, but they don't!

RazvanN7 commented:
@CyberShadow As I see it, there are 2 classes of skipOver methods: those that skip over multiple range elements and those that skip over the first element. This PR enhances the first category.

The 2 cases you are taking about are likely to appear in the real world and skipOver should manage them both. I don't see any inconsistency here.

andralex commented:
@CyberShadow it's a point, but the usefulness in parsing is there.

* `s.skipOver(' ')` ---> "skip over one space"
* `s.skipOver(c => c == ' ')` ---> "skip over any number of spaces"
* `s.skipOver(c => c.among(' ', '\t', '\r', '\n')` ----> "skip all whitespace"

CyberShadow commented:
It would be less confusing if there was a different overload set for functions that skip over multiple instances of the pattern.

CyberShadow commented:
* `s.skipOver(' ')` ---> "skip over one space"
* `s.skipOver(c => c.among(' ', '\t', '\r', '\n')` ----> "skip over one whitespace character"

* `s.skipWhile(' ')` ---> "skip over spaces"
* `s.skipWhile(c => c.among(' ', '\t', '\r', '\n')` ----> "skip all whitespace"

The return types should be different. The version that skips multiple prefixes ought to return a `size_t`.

andralex commented:
@CyberShadow that's awfully nice, @RazvanN7 how about implementing that real quick?

CyberShadow commented:
@andralex @RazvanN7 Found the previous discussion: https://github.com/dlang/phobos/pull/5497#issuecomment-309599156

Looks like we already have some primitives in Phobos to do some of these things. CC @wilzbach
